### PR TITLE
feat(lua): vim.keycode

### PR DESCRIPTION
**Pull request by ii14**
Using nvim_replace_termcodes is too verbose, add vim.keycode for translating keycodes.

### CODE REVIEW

**Comment by clason** [comment]
`vim.tr` is bad discoverability, I think. If it needs to be of minimal length (which I doubt), then `vim.t`. But I'd prefer something more following the standard naming guidelines, as people will do `local t = vim.<whatever>` anyway in order to do `t'<bs>'`.

**Comment by ii14** [comment]
`vim.replace_keycodes` then?

File: runtime/doc/lua.txt [hunk 1]
```diff
@@ -1541,6 +1541,25 @@ schedule_wrap({cb})                                      *vim.schedule_wrap()*
...
+
+    Parameters: ~
+      • {str}   string String to be converted.
+      • {opts}  table|nil Optional arguments:
+                • do_lt: boolean|nil (default: true) Also translate <lt>.
+                  Ignored if `special` is false.
+                • special: boolean|nil (default: true) Replace |keycodes|,
+                  e.g. <CR> becomes a "\r" char.
+                • from_part: boolean|nil (default: true) Legacy Vim parameter.
+                  Usually true.
```
Comment (zeertzjq) [inline runtime/doc/lua.txt:1555]: I'd recommend removing this from docs, because the effect of this argument is very confusing.

**Comment by lewis6991** [comment]
Or `vim.termcode`?

**Comment by ii14** [comment]
The term `termcode` doesn't appear anywhere in the documentation other than in `nvim_replace_termcodes`, and I'm not even 100% sure what that actually means. I'd prefer `vim.keycode` or `vim.keycodes`

**Comment by lewis6991** [comment]
Makes sense. Either way, I don't think this is used enough to justify a 2 letter name, so something like `vim.keycode` works well.

**Comment by clason** [comment]
Maybe add the suggestion to the docs:
```lua
local k = vim.keycode
vim.g.mapleader = k'<bs>'
```
?

**Comment by haolian9** [comment]
sorry if i'm too rude to comment here. is it worth to cache the result?
i have a similar wrapper on nvim_replace_termcodes with result caching, but never take a deep look of it to see if it is worth or not.

File: runtime/doc/lua.txt [hunk 2]
```diff
@@ -1404,6 +1404,25 @@ inspect({object}, {options})                                   *vim.inspect()*
...
+    Translate keycodes.
+
+    Parameters: ~
+      • {str}   string String to be converted.
+      • {opts}  table|nil Optional arguments:
+                • do_lt: boolean|nil (default: true) Also translate <lt>.
+                  Ignored if `special` is false.
+                • special: boolean|nil (default: true) Replace |keycodes|,
+                  e.g. <CR> becomes a "\r" char.
+                • from_part: boolean|nil (default: true) Legacy Vim parameter.
```
Comment (lewis6991) [inline runtime/doc/lua.txt:1417]: Is it even worth including this? This just seems like noise.

**Review by lewis6991 — COMMENTED** [review]
I think there may be an argument in simply adding a new API function and deprecating the old one. There has been some discussion on allowing the opts parameter to be optional in Lua. I can't remember the exact details. @bfredl do you remember what was agreed?

**Comment by bfredl** [comment]
Not sure I would add a new API function just for this. 

TBH I was thinking we should just add 

    function vim.keycode(x) 
      return vim.api.nvim_replace_termcodes(x, true, true, true) 
    end

Not everything needs to be batshit overconfigurable. just always use `<lt>` instead of provide the option of using it or not, etc. (If you need to change things for some reason, just use nvim_replace_termcodes directly..)

**Review by clason — APPROVED** [review]

File: runtime/lua/vim/_editor.lua [hunk 3]
```diff
@@ -811,6 +811,21 @@ function vim.print(...)
...
+--- Example:
+--- <pre>lua
+---   local k = vim.keycode
+---   vim.g.mapleader = k'<bs>'
+--- </pre>
+--- @param str string String to be converted.
+--- @return string
+--- @see |nvim_replace_termcodes()|
+function vim.keycode(str)
+  vim.validate({ str = { str, 's' } })
```
Comment (lewis6991) [inline runtime/lua/vim/_editor.lua:825]: Won't `nvim_replace_termcodes` do this for us?

**Review by lewis6991 — APPROVED** [review]

**Pull request merged**